feat(coderd): notify when workspace is marked as dormant

BrunoQuaresma opened the pull request:
Related to https://github.com/coder/team-coconut/issues/7

BrunoQuaresma committed:
feat(coderd): notify when workspace is marked as dormant

BrunoQuaresma committed:
Fix role

dannykopping commented on coderd/workspaces.go:
```diff
@@ -952,6 +953,29 @@ func (api *API) putWorkspaceDormant(rw http.ResponseWriter, r *http.Request) {
+			dormancy.WorkspaceDormantNotification{
+				Workspace:   workspace,
+				InitiatedBy: initiator.Username,
+				Reason:      "requested by user",
```
Is this API only ever called by user-initiated actions? I had a brief look and it seems to be, but maybe you have more context. If that's the case, we don't need this constant value - we can just always say "requested by user" in the template.

BrunoQuaresma replied:
I need to implement this notification in two places: the lifecycle executor and the scheduler.

dannykopping commented on coderd/dormancy/notifications.go:
```diff
@@ -0,0 +1,43 @@
+package dormancy
```
Since dormancy is an enterprise-only feature, you should consider moving this to the existing `dormancy` package in `enterprise/coderd/dormancy`.

dannykopping commented on coderd/workspaces_test.go:
```diff
@@ -3504,3 +3506,140 @@ func TestWorkspaceUsageTracking(t *testing.T) {
+			)
+
+			ctx, cancel := context.WithTimeout(context.Background(), testutil.WaitLong)
+			defer cancel()
```
```suggestion
			t.Cleanup(cancel)
```

dannykopping commented on coderd/workspaces_test.go:
```diff
@@ -3504,3 +3506,140 @@ func TestWorkspaceUsageTracking(t *testing.T) {
+	})
+}
+
+type fakeNotificationEnqueuer struct {
+	mu   sync.Mutex
+	sent []*notification
+}
+
+type notification struct {
+	userID, templateID uuid.UUID
+	labels             map[string]string
+	createdBy          string
+	targets            []uuid.UUID
+}
+
+func (f *fakeNotificationEnqueuer) Enqueue(_ context.Context, userID, templateID uuid.UUID, labels map[string]string, createdBy string, targets ...uuid.UUID) (*uuid.UUID, error) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	f.sent = append(f.sent, &notification{
+		userID:     userID,
+		templateID: templateID,
+		labels:     labels,
+		createdBy:  createdBy,
+		targets:    targets,
+	})
+
+	id := uuid.New()
+	return &id, nil
+}
```
Rather than copying this from `coderd/provisionerdserver/provisionerdserver_test.go`, let's move this into the `notifications` package under test utilities.

dannykopping commented on coderd/workspaces.go:
```diff
@@ -952,6 +953,29 @@ func (api *API) putWorkspaceDormant(rw http.ResponseWriter, r *http.Request) {
+			})
+			return
+		}
+		dormancy.NotifyWorkspaceDormant(
```
Is this the only place this will be called? If so, I don't see the need for the helper method or the struct.

BrunoQuaresma replied:
Nops, we are going to call them in the lifecycle executor and the scheduler.

dannykopping commented on coderd/database/migrations/000221_notifications.up.sql:
```diff
@@ -50,16 +50,35 @@ CREATE TABLE notification_messages
+              "url": "{{ base_url }}/templates"
+          }
+      ]'::jsonb),
+	-- Workspace Marked as Dormant
```
You should move this out into its own migration; customers who have already run this migration will not get the new template.

dannykopping commented on coderd/workspaces_test.go:
```diff
@@ -3504,3 +3506,140 @@ func TestWorkspaceUsageTracking(t *testing.T) {
+	return &id, nil
+}
+
+func (f *fakeNotificationEnqueuer) Clear() {
```
👍 nice

dannykopping commented on coderd/workspaces_test.go:
```diff
@@ -3504,3 +3506,140 @@ func TestWorkspaceUsageTracking(t *testing.T) {
 	})
 }
+
+func TestNotifications(t *testing.T) {
```
Nice tests 👍 
Nit: you could add some more comments to separate `GIVEN`/`WHEN`/`THEN` and explain what you're doing to make them more readable, but they're quite simple and clean already.

dannykopping reviewed (commented):
Good work, left a few thoughts

BrunoQuaresma committed:
Apply Danny suggestions

BrunoQuaresma committed:
Notify dormant workspace on lifecycle executor

BrunoQuaresma committed:
Notify dormancy on template schedule

dannykopping commented on coderd/database/migrations/000223_dormancy_notification_template.up.sql:
```diff
@@ -0,0 +1,16 @@
+-- TODO: autogenerate constants which reference the UUIDs
```
Are you up-to-date with `main`? We're already up to `225`: https://github.com/coder/coder/tree/main/coderd/database/migrations

dannykopping commented on coderd/workspaces.go:
```diff
@@ -957,21 +957,18 @@ func (api *API) putWorkspaceDormant(rw http.ResponseWriter, r *http.Request) {
-				Message: "Internal error fetching user.",
-				Detail:  err.Error(),
-			})
+			api.Logger.Warn(ctx, "failed to get the initiator by id", slog.Error(err))
```
We shouldn't return in this scenario

BrunoQuaresma replied:
💯 🤦

dannykopping commented on coderd/autobuild/lifecycle_executor.go:
```diff
@@ -215,6 +220,18 @@ func (e *Executor) runOnce(t time.Time) Stats {
 							},
 						})
 
+						dormancy.NotifyWorkspaceDormant(
```
I think this should go after the error is checked to be `nil`.

dannykopping commented on enterprise/coderd/schedule/template.go:
```diff
@@ -159,14 +166,27 @@ func (s *EnterpriseTemplateScheduleStore) Set(ctx context.Context, db database.S
 		// disabling it (by passing 0), then we want to delete nullify the deleting_at
 		// fields of all the template workspaces.
-		err = tx.UpdateWorkspacesDormantDeletingAtByTemplateID(ctx, database.UpdateWorkspacesDormantDeletingAtByTemplateIDParams{
+		dormantWorspaces, err := tx.UpdateWorkspacesDormantDeletingAtByTemplateID(ctx, database.UpdateWorkspacesDormantDeletingAtByTemplateIDParams{
```
We'll have to do this outside of the transaction. If the transaction fails and needs to be rolled back, then the notification will still be sent and will indicate an action which didn't complete.

dannykopping commented on enterprise/coderd/schedule/template_test.go:
```diff
@@ -297,7 +298,7 @@ func TestTemplateUpdateBuildDeadlines(t *testing.T) {
 				TimeTilDormant:           0,
 				TimeTilDormantAutoDelete: 0,
-			})
+			}, nil, slog.Logger{})
```
Please pass a real logger here in case something fails in the tests.

```go
slogtest.Make(t, &slogtest.Options{IgnoreErrors: true}).Leveled(slog.LevelDebug)
```

dannykopping commented on coderd/database/dbauthz/dbauthz.go:
```diff
@@ -3526,12 +3526,15 @@ func (q *querier) UpdateWorkspaceTTL(ctx context.Context, arg database.UpdateWor
-	fetch := func(ctx context.Context, arg database.UpdateWorkspacesDormantDeletingAtByTemplateIDParams) (database.Template, error) {
-		return q.db.GetTemplateByID(ctx, arg.TemplateID)
+func (q *querier) UpdateWorkspacesDormantDeletingAtByTemplateID(ctx context.Context, arg database.UpdateWorkspacesDormantDeletingAtByTemplateIDParams) ([]database.Workspace, error) {
+	template, err := q.db.GetTemplateByID(ctx, arg.TemplateID)
```
Be careful that the semantics didn't change here since you're moving away from `fetchAndExec`.

BrunoQuaresma replied:
What do you mean? 🤔

dannykopping replied:
Previously the function was using `fetchAndExec` but now you've replaced that, so just take care that this didn't change the behaviour of this function.

BrunoQuaresma replied:
Now it is returning the affected workspaces, which I think is okay and is related to this method signature. Wdyt?

BrunoQuaresma committed:
Apply Danny review suggestions

BrunoQuaresma committed:
Fix mismatch usage

BrunoQuaresma committed:
Resolve migration conflict

BrunoQuaresma committed:
Merge branch 'main' of https://github.com/coder/coder into bq/implement-notifications

BrunoQuaresma commented on enterprise/coderd/dormancy/notifications.go:
```diff
@@ -0,0 +1,43 @@
+		notification.Workspace.TemplateID,
+		notification.Workspace.OrganizationID,
+	); err != nil {
+		logger.Warn(ctx, "failed to notify of workspace marked as dormant", slog.Error(err))
```
That is completely true. I don't know why I didn't see that before.

BrunoQuaresma committed:
Return error instead of receiving the logger

BrunoQuaresma committed:
Improve verbiage

dannykopping commented on coderd/database/migrations/000226_dormancy_notification_template.up.sql:
```diff
@@ -0,0 +1,9 @@
+INSERT INTO notification_templates (id, name, title_template, body_template, "group", actions)
+VALUES ('123e4567-e89b-12d3-a456-426614174000', 'Workspace Marked as Dormant', E'Workspace "{{.Labels.name}}" marked as dormant',
+        E'Hi {{.UserName}}\n\nYour workspace **{{.Labels.name}}** has been marked as dormant.\nThe specified reason was "**{{.Labels.reason}}{{ if .Labels.initiator }} ({{ .Labels.initiator }}){{end}}**',
```
I think we'll need to add some detail here about what "dormant" means; perhaps a link to the docs and how to fix it.

dannykopping commented on coderd/schedule/mock.go:
```diff
@@ -5,12 +5,15 @@ import (
 type MockTemplateScheduleStore struct {
 	GetFn func(ctx context.Context, db database.Store, templateID uuid.UUID) (TemplateScheduleOptions, error)
-	SetFn func(ctx context.Context, db database.Store, template database.Template, options TemplateScheduleOptions) (database.Template, error)
+	SetFn func(ctx context.Context, db database.Store, template database.Template, options TemplateScheduleOptions, enqueuer notifications.Enqueuer, logger slog.Logger) (database.Template, error)
```
Let's remove these loggers since they're not needed anymore.

dannykopping committed:
Merge branch 'main' of github.com:/coder/coder into bq/implement-notifications

dannykopping committed:
Merge branch 'main' of github.com:/coder/coder into bq/implement-notifications

dannykopping committed:
Possible implementation simplification

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping commented:
@BrunoQuaresma I tried to move this along while you were out, but I didn't have much time.
I added https://github.com/coder/coder/pull/13868/commits/467a797edcd927d66c47f0ef44b52f0bee22f05e which simplifies the implementation, although I have not had time to test it thoroughly; feel free to revert if it causes issues.

I think we should pass the enqueuer during the instantiation phase of the various areas, rather than passing the enqueuer/logger at the last moment of responsibility like you were doing before.

I fixed the migration and added some detail to the notification.

BrunoQuaresma committed:
Merge branch 'main' of https://github.com/coder/coder into bq/implement-notifications

BrunoQuaresma committed:
Apply fmt

BrunoQuaresma committed:
Notify after executor is done

dannykopping commented on coderd/workspaces.go:
```diff
@@ -950,6 +951,28 @@ func (api *API) putWorkspaceDormant(rw http.ResponseWriter, r *http.Request) {
+	if req.Dormant && apiKey.UserID != workspace.OwnerID {
+		initiator, err := api.Database.GetUserByID(ctx, apiKey.UserID)
+		if err != nil {
+			api.Logger.Warn(ctx, "failed to fetch the user that marked the workspace as dormant", slog.Error(err))
```
Let's log the workspace name as well, otherwise this warning isn't very helpful.

dannykopping reviewed (approved):
LGTM! This looks good @BrunoQuaresma 

Have you tested with a real workspace, or just with unit tests?

dannykopping committed:
Revert refactoring mistake

Signed-off-by: Danny Kopping <danny@coder.com>

BrunoQuaresma commented:
> Have you tested with a real workspace, or just with unit tests?

Just unit tests, but I would like to know how to QA it in a real scenario.

BrunoQuaresma committed:
Add workspace name to the log

BrunoQuaresma committed:
Set a fake enqueuer on coderdtest options

BrunoQuaresma committed:
Merge branch 'bq/implement-notifications' of https://github.com/coder/coder into bq/implement-notifications

dannykopping commented:
> @dannykopping just to make it clear, this PR is missing tests in the lifecycle executor and templates schedule, but I'm trying to first fix the current broken tests.

Shout if you need a hand.

BrunoQuaresma committed:
Fix migration

BrunoQuaresma committed:
Fix migration sql

BrunoQuaresma committed:
Merge branch 'main' of https://github.com/coder/coder into bq/implement-notifications

BrunoQuaresma committed:
Fix typo

BrunoQuaresma committed:
Fix migration number

BrunoQuaresma committed:
Merge branch 'main' of https://github.com/coder/coder into bq/implement-notifications

BrunoQuaresma commented on coderd/autobuild/lifecycle_executor.go:
```diff
@@ -228,6 +233,21 @@ func (e *Executor) runOnce(t time.Time) Stats {
+								Workspace: ws,
+								Initiator: "system",
+								Reason:    "breached the template's threshold for inactivity",
+								CreatedBy: "lifecycleexecutor",
```
I saw @dannykopping use it in the notification, so I just copied it, but I think it is helpful for searching.

BrunoQuaresma commented on coderd/database/migrations/000227_dormancy_notification_template.up.sql:
```diff
@@ -0,0 +1,13 @@
+INSERT INTO notification_templates (id, name, title_template, body_template, "group", actions)
+VALUES ('123e4567-e89b-12d3-a456-426614174000', 'Workspace Marked as Dormant', E'Workspace "{{.Labels.name}}" marked as dormant',
+        E'Hi {{.UserName}}\n\n' ||
```
It was @dannykopping's refactoring, but I liked it more. It makes it easier to read, IMO.

dannykopping replied:
`||` is just for concatenation; it's just for formatting in this file. The newlines are still necessary for output formatting.

mtojek replied:
I see, so it's a personal preference. Let's leave it then.

BrunoQuaresma commented on coderd/database/migrations/000227_dormancy_notification_template.up.sql:
```diff
@@ -0,0 +1,13 @@
+VALUES ('123e4567-e89b-12d3-a456-426614174000', 'Workspace Marked as Dormant', E'Workspace "{{.Labels.name}}" marked as dormant',
+        E'Hi {{.UserName}}\n\n' ||
+        E'Your workspace **{{.Labels.name}}** has been marked as **dormant**.\n' ||
+        E'The specified reason was "**{{.Labels.reason}}{{ if .Labels.initiator }} (initiated by: {{ .Labels.initiator }}){{end}}**\n\n' ||
```
Why?

mtojek replied:
That's why I'm asking :) In the template you have:

`{{ if .Labels.initiator }} {{ .Labels.initiator }} {{end}}`, which suggests that _Initiator_ is optional and may not be defined. I think you can convert it to `{{ .Labels.initiator }}`. Does it sound about right, @dannykopping?

dannykopping replied:
Ah, I interpreted your comment differently, and I suspect Bruno did too. It sounded like you were asking if we can drop the `initiator` field for some reason.

Yes, if the label is never blank then we don't need the conditional.

mtojek replied:
Sorry for the shortcut thinking!

BrunoQuaresma committed:
Apply Marcin comments

BrunoQuaresma committed:
Fix lint

dannykopping commented on coderd/autobuild/lifecycle_executor.go:
```diff
@@ -331,6 +325,16 @@ func (e *Executor) runOnce(t time.Time) Stats {
+						*dormantNotification,
+					)
+					if err != nil {
+						e.log.Warn(e.ctx, "failed to notify of workspace "+dormantNotification.Workspace.Name+" marked as dormant", slog.Error(err))
```
Log messages should generally be static, and all variability should be in fields.
I'd also recommend logging the workspace id not name since it's universally unique.

```suggestion
						e.log.Warn(e.ctx, "failed to notify of workspace marked as dormant", slog.Error(err), slog.F("workspace", dormantNotification.Workspace.Id))
```

dannykopping commented on coderd/schedule/mock.go:
```diff
@@ -23,12 +23,12 @@ func (m MockTemplateScheduleStore) Get(ctx context.Context, db database.Store, t
 	return NewAGPLTemplateScheduleStore().Get(ctx, db, templateID)
 }
 
-func (m MockTemplateScheduleStore) Set(ctx context.Context, db database.Store, template database.Template, opts TemplateScheduleOptions) (database.Template, error) {
```
Please revert this file; this was a bad refactor from my side which I fixed, but it seems you must've had a conflict and restored the changes.

BrunoQuaresma replied:
This was reverted already.

dannykopping commented on coderd/workspaces.go:
```diff
@@ -955,7 +955,7 @@ func (api *API) putWorkspaceDormant(rw http.ResponseWriter, r *http.Request) {
 		initiator, err := api.Database.GetUserByID(ctx, apiKey.UserID)
 		if err != nil {
-			api.Logger.Warn(ctx, "failed to fetch the user that marked the workspace as dormant", slog.Error(err))
+			api.Logger.Warn(ctx, "failed to fetch the user that marked the workspace "+workspace.Name+" as dormant", slog.Error(err))
```
Same comment as the previous case

Commit:
Update coderd/autobuild/lifecycle_executor.go

Co-authored-by: Danny Kopping <danny@coder.com>

BrunoQuaresma committed:
Apply dannys comment

mtojek commented:
@BrunoQuaresma is it still a draft PR or a regular one?

dannykopping commented:
> @BrunoQuaresma is it still a draft PR or a regular one?

I made it draft so that we don't merge accidentally before testing it fully, since it has approvals already.

BrunoQuaresma committed:
Simplify dormancy template

BrunoQuaresma committed:
Merge branch 'main' of https://github.com/coder/coder into bq/implement-notifications

BrunoQuaresma committed:
Add test to verify dormancy in lifecycle executor

BrunoQuaresma committed:
Add placeholder

sreya committed:
make test pass

mtojek committed:
fix: lint

BrunoQuaresma committed:
Merge branch 'main' of https://github.com/coder/coder into bq/implement-notifications

BrunoQuaresma committed:
Add test to verify notification in lifecycle executor

BrunoQuaresma committed:
Add notification for marked as deletion

BrunoQuaresma commented:
@dannykopping @mtojek The biggest changes since the last review:
- Added a test for the dormant notification in the lifecycle executor
- Created a different template for "Marked for Deletion" since it is a bit different from the "Marked as Dormant" use case
- Added tests to verify if notifications are sent when the dormant TTL is changed and if affected workspaces are getting notified

dannykopping commented on coderd/database/migrations/000230_marked_for_deletion_notification_template.up sql:
```diff
@@ -0,0 +1,22 @@
+        "group",
+        actions
+    )
+VALUES (
+        '51ce2fdf-c9ca-4be1-8d70-628674f9bc42',
+        'Workspace Marked for Deletion',
+        E'Workspace "{{.Labels.name}}" marked for deletion',
+        E'Hi {{.UserName}}\n\n' || E'Your workspace **{{.Labels.name}}** has been marked for **deletion** after {{.Labels.dormancyHours}} hours of dormancy.\n' || E'The specified reason was "**{{.Labels.reason}} (initiated by: {{ .Labels.initiator }}){{end}}**\n\n' || E'Dormancy refers to a workspace being unused for a defined length of time, and after it exceeds {{.Labels.dormancyHours}} hours o ...
```
Nit: I'm not sure `(initiated by: autobuild)` is going to be useful to folks.
I think we can drop it from the messaging. Initiator is useful in the `workspace deleted` notification because it can be initiated by a human, but this isn't so it won't help the receiver.

mtojek replied:
+1 to drop this

dannykopping commented on enterprise/coderd/schedule/template.go:
```diff
@@ -205,19 +205,18 @@ func (s *EnterpriseTemplateScheduleStore) Set(ctx context.Context, db database.S
 		)
 		if err != nil {
-			s.logger.Warn(ctx, "failed to notify of workspace marked as dormant", slog.Error(err))
+			s.logger.Warn(ctx, "failed to notify of workspace marked for deletion", slog.Error(err))
```
Which workspace?

dannykopping commented on enterprise/coderd/schedule/template_test.go:
```diff
@@ -606,6 +606,95 @@ func TestTemplateUpdateBuildDeadlinesSkip(t *testing.T) {
+					"foo": "bar",
+				},
+			})
+			timeTilDormant  = time.Minute * 2
```
Ditto

dannykopping commented on enterprise/coderd/schedule/template_test.go:
```diff
@@ -606,6 +606,95 @@ func TestTemplateUpdateBuildDeadlinesSkip(t *testing.T) {
+		templateScheduleStore := schedule.NewEnterpriseTemplateScheduleStore(userQuietHoursStorePtr, &notifyEnq, logger)
+		templateScheduleStore.TimeNowFn = time.Now
+
+		// Update dormancy TTL for a lower value
```
Nit: explain the _why_ not the _what_; the what is right there, but the reasoning for it is what's important.

dannykopping commented on enterprise/coderd/schedule/template_test.go:
```diff
@@ -606,6 +606,95 @@ func TestTemplateUpdateBuildDeadlinesSkip(t *testing.T) {
+		require.NoError(t, err)
+
+		// We should expect two notifications. One for each dormant workspace.
+		require.Len(t, notifyEnq.Sent, len(dormantWorkspaces))
```
Can we be more specific here?
Which notification is being received? What are its labels? etc.
If you had this you would've caught that your new notification was never being sent.

mtojek replied:
+1

mtojek commented on coderd/database/migrations/000230_marked_for_deletion_notification_template.up sql:
```diff
@@ -0,0 +1,22 @@
+INSERT INTO
```
Interesting that our tests didn't catch it.

mtojek commented on coderd/autobuild/lifecycle_executor.go:
```diff
@@ -332,8 +332,10 @@ func (e *Executor) runOnce(t time.Time) Stats {
+						log.Warn(e.ctx, "failed to notify of workspace marked as dormant", slog.Error(err), slog.F("workspace_id", dormantNotification.Workspace.ID))
 					}
+				} else {
+					log.Warn(e.ctx, "no dormant notification to send", slog.F("workspace_id", wsID))
```
👍 for removing this

BrunoQuaresma replied:
Idk why I put it there, thanks for noticing it.

mtojek commented on coderd/autobuild/lifecycle_executor_test.go:
```diff
@@ -1062,6 +1062,68 @@ func TestExecutorInactiveWorkspace(t *testing.T) {
+
+		// Check that the workspace is dormant
+		workspace = coderdtest.MustWorkspace(t, client, workspace.ID)
+		require.NotNil(t, workspace.DormantAt)
```
.. unless it is flaky, then remove this check

BrunoQuaresma replied:
Since it is a background job, I think it is hard to find the exact time. Here is a diff between the value and the expectation:
```
expected: *time.Time(time.Date(2024, time.July, 24, 13, 37, 57, 414817000, time.UTC))
actual  : time.Time(time.Date(2024, time.July, 24, 13, 38, 56, 603858000, time.UTC))
```
I also don't think knowing if the dormant time is correct is helpful for this particular test.

mtojek reviewed (commented):
I seconded a few Danny's comments.

@BrunoQuaresma If you feel a need to clarify some comments, please sync with @dannykopping. Let's not elongate this task anymore, it is already a leftover from the last sprint.

BrunoQuaresma commented on coderd/database/migrations/000229_dormancy_notification_template.up.sql:
```diff
@@ -1,13 +1,22 @@
+        '0ea69165-ec14-4314-91f1-69566ac3c5a0',
+        'Workspace Marked as Dormant',
+        E'Workspace "{{.Labels.name}}" marked as dormant',
+        E'Hi {{.UserName}}\n\n' || E'Your workspace **{{.Labels.name}}** has been marked as **dormant**.\n' || E'The specified reason was "**{{.Labels.reason}} (initiated by: {{ .Labels.initiator }}){{end}}**\n\n' || E'Dormancy refers to a workspace being unused for a defined length of time, and after it exceeds {{.Labels.dormancyHours}} hours of dormancy it will be deleted.\n' || E'To prevent yo ...
```
What do you think about:
- `hours of dormancy it will be deleted` to `hours of dormancy might be deleted`
- `To prevent your workspace from being deleted, simply use it as normal.` to `To activate your workspace again, simply use it as normal.`

BrunoQuaresma replied:
I'm moving forward with the above but we can change it later.

BrunoQuaresma committed:
Apply Dannys review comments

BrunoQuaresma committed:
Fix SQL

BrunoQuaresma committed:
Rollback dormant at test

BrunoQuaresma committed:
Fix template

stirby commented on coderd/database/migrations/000229_dormancy_notification_template.up.sql:
```diff
@@ -0,0 +1,35 @@
+        '0ea69165-ec14-4314-91f1-69566ac3c5a0',
+        'Workspace Marked as Dormant',
+        E'Workspace "{{.Labels.name}}" marked as dormant',
+        E'Hi {{.UserName}}\n\n' || E'Your workspace **{{.Labels.name}}** has been marked as **dormant**.\n' || E'The specified reason was "**{{.Labels.reason}} (initiated by: {{ .Labels.initiator }}){{end}}**\n\n' || E'Dormancy refers to a workspace being unused for a defined length of time, and after it exceeds {{.Labels.dormancyHours}} hours of dormancy might be deleted.\n' || E'To activate you ...
```
The wording is a little weird here. Can we change it and add a link to the dormancy docs:

I don't think we need to include the initiator, just improve the reasons (see end of message).

> Hi `{user-name}`, your workspace `{ws-name}` has been marked [dormant](https://coder.com/docs/templates/schedule#dormancy-threshold-enterprise) because of `{reason}`.

Append if dormancy cleanup is enabled and `dormancy_hours` < 24:
> Dormant workspaces are [automatically deleted](https://coder.com/docs/templates/schedule#dormancy-auto-deletion-enterprise) after `{dormancy_hours}` hours of inactivity.

Append if dormancy cleanup is enabled, and `dormancy_hours` > 24:
> Dormant workspaces are [automatically deleted](https://coder.com/docs/templates/schedule#dormancy-auto-deletion-enterprise) after `{dormancy_hours // 24}` days of inactivity.

> To prevent deletion, use your workspace with the link below.

---

The reasons also fit into differing grammatic context:
- Lifecycle executor: "breached the template's threshold for inactivity"
- API: "requested by user"
- Template: "template updated to new dormancy policy"

If we use these, they'd fit the above messages better.
- Lifcycle: "prolonged inactivity, exceeding the dormancy threshold"
- API: "a user request"
- Template: "an update to the template's dormancy"

BrunoQuaresma replied:
Going to refactor the templates in the next PR to avoid further postponement.
